File: python-version/statistics_calculator.py

"""
Program: Statistics Calculator (Python Version)
This program asks the user to enter a list of integers and then calculates
the Mean, Median, and Mode using an Object-Oriented Programming (OOP) approach.
"""

from collections import Counter

class StatisticsCalculator:
    """
    This class encapsulates all functionality for calculating
    mean, median, and mode from a list of integers.
    """

    def __init__(self, numbers):
        """
        Constructor to initialize the list of integers.
        :param numbers: List of integers.
        """
        self.numbers = numbers

    def calculate_mean(self):
        """
        Calculates the mean (average) of the list of integers.
        """
        total = sum(self.numbers)
        count = len(self.numbers)
        mean = total / count
        return mean

    def calculate_median(self):
        """
        Calculates the median (middle value) of the list.
        """
        sorted_numbers = sorted(self.numbers)
        n = len(sorted_numbers)

        if n % 2 == 1:
            median = sorted_numbers[n // 2]
        else:
            mid1 = sorted_numbers[n // 2 - 1]
            mid2 = sorted_numbers[n // 2]
            median = (mid1 + mid2) / 2

        return median

    def calculate_mode(self):
        """
        Calculates the mode (most frequent value(s)) in the list.
        """
        counts = Counter(self.numbers)
        max_freq = max(counts.values())
        mode = [num for num, freq in counts.items() if freq == max_freq]
        return mode


def get_user_input():
    """
    Prompts the user to enter integers separated by space,
    and returns them as a list of integers.
    """
    while True:
        try:
            raw_input = input("Enter integers separated by space: ")
            numbers = list(map(int, raw_input.strip().split()))
            if len(numbers) == 0:
                print("Please enter at least one number.")
                continue
            return numbers
        except ValueError:
            print("Invalid input. Please enter only integers.")


# Main logic
if __name__ == "__main__":
    # Ask user for input
    user_numbers = get_user_input()

    # Create calculator instance
    calc = StatisticsCalculator(user_numbers)

    # Print results
    print("\nInput Numbers:", user_numbers)
    print("Mean:", calc.calculate_mean())
    print("Median:", calc.calculate_median())
    print("Mode:", calc.calculate_mode())

File: ocaml-version/statistics_calculator.ml
(*
  Program: Statistics Calculator (OCaml Version with User Input)
  This version allows the user to input a list of integers.
  It calculates Mean, Median, and Mode using functional style.
*)

(* Helper to split a string into a list of integers *)
let parse_input input =
  input
  |> String.split_on_char ' '
  |> List.filter (fun s -> s <> "")
  |> List.map int_of_string

(* Calculate mean *)
let calculate_mean lst =
  let sum = List.fold_left ( + ) 0 lst in
  let count = List.length lst in
  float_of_int sum /. float_of_int count

(* Sort a list *)
let sort_list lst = List.sort compare lst

(* Calculate median *)
let calculate_median lst =
  let sorted = sort_list lst in
  let len = List.length sorted in
  if len mod 2 = 1 then
    float_of_int (List.nth sorted (len / 2))
  else
    let mid1 = List.nth sorted (len / 2 - 1) in
    let mid2 = List.nth sorted (len / 2) in
    float_of_int (mid1 + mid2) /. 2.0

(* Count frequencies *)
let count_frequencies lst =
  let update_count map num =
    if List.mem_assoc num map then
      (num, (List.assoc num map) + 1) :: (List.remove_assoc num map)
    else
      (num, 1) :: map
  in
  List.fold_left update_count [] lst

(* Calculate mode *)
let calculate_mode lst =
  let freq_list = count_frequencies lst in
  let max_freq = List.fold_left (fun acc (_, freq) -> max acc freq) 0 freq_list in
  List.fold_left (fun acc (num, freq) ->
    if freq = max_freq then num :: acc else acc
  ) [] freq_list
  |> sort_list

(* MAIN *)
let () =
  print_endline "Enter integers separated by space:";
  let input = read_line () in
  let numbers =
    try parse_input input
    with Failure _ ->
      print_endline "Invalid input. Please enter only integers.";
      exit 1
  in

  Printf.printf "Input Numbers: ";
  List.iter (fun x -> Printf.printf "%d " x) numbers;
  Printf.printf "\n";

  let mean = calculate_mean numbers in
  let median = calculate_median numbers in
  let mode = calculate_mode numbers in

  Printf.printf "Mean: %.2f\n" mean;
  Printf.printf "Median: %.2f\n" median;
  Printf.printf "Mode: ";
  List.iter (fun x -> Printf.printf "%d " x) mode;
  Printf.printf "\n"